main.go

```diff
@@ -131,59 +142,87 @@ func main() {
 		faasClient:             faasClient,
 	}
 
-	if !operator {
-		log.Println("Starting controller")
-		runController(setup)
-	} else {
+	if operator {
 		log.Println("Starting operator")
 		runOperator(setup, config)
+	} else {
+		log.Println("Starting controller")
+		runController(setup)
 	}
 }
 
-// runController runs the faas-netes imperative controller
-func runController(setup serverSetup) {
-	// pull out the required config and clients fromthe setup, this is largely a
-	// leftover from refactoring the setup to a shared step and keeping the function
-	// signature readable
-	config := setup.config
-	kubeClient := setup.kubeClient
-	factory := setup.functionFactory
+type customInformers struct {
+	EndpointsInformer  v1core.EndpointsInformer
+	DeploymentInformer v1apps.DeploymentInformer
+	FunctionsInformer  v1.FunctionInformer
+}
+
+func startInformers(setup serverSetup, stopCh <-chan struct{}, operator bool) customInformers {
 	kubeInformerFactory := setup.kubeInformerFactory
 	faasInformerFactory := setup.faasInformerFactory
 
-	// set up signals so we handle the first shutdown signal gracefully
-	stopCh := signals.SetupSignalHandler()
-
-	endpointsInformer := kubeInformerFactory.Core().V1().Endpoints()
+	var functions v1.FunctionInformer
+	if operator {
+		go faasInformerFactory.Start(stopCh)
+		glog.Infof("Waiting for cache sync for functions")
+		functions = faasInformerFactory.Openfaas().V1().Functions()
 
-	deploymentLister := kubeInformerFactory.Apps().V1().
-		Deployments().Lister()
+		go functions.Informer().Run(stopCh)
+		if ok := cache.WaitForNamedCacheSync("faas-netes", stopCh, functions.Informer().HasSynced); !ok {
+			log.Fatalf("failed to wait for cache to sync")
+		}
+	}
 
-	go faasInformerFactory.Start(stopCh)
 	go kubeInformerFactory.Start(stopCh)
+	glog.Infof("Waiting for cache sync for Deployments")
+	deployments := kubeInformerFactory.Apps().V1().Deployments()
+	go deployments.Informer().Run(stopCh)
+	if ok := cache.WaitForNamedCacheSync("faas-netes", stopCh, deployments.Informer().HasSynced); !ok {
```

Comment:
Nit: Not sure what is preferable but you might not need the `glog` statement on L177 if you changed the string argument from `faas-netes` to `Deployments` or something else as the logging is taken care of [here](https://github.com/kubernetes/client-go/blob/758467711e075d6fd3d31abcaf6e2e1eb51ef3d4/tools/cache/shared_informer.go#L240)